### ISSUE TITLE
RecyclerView disappearing while loading images

### ISSUE THREAD

**Issue by summers314** [bumptech/glide#1449]
Glide 3.7.0

Android - Two Way View (Recycler View)
https://github.com/lucasr/twoway-view

Moto G

While using Glide to load Images. The first time the images are loaded (even when they are saved in cache). The view starts to hang before disappearing. (Note: App doesn't crash. The view simply disappears). 

I've done a side by side comparison of Picasso and Glide. While it works properly with Picasso, the load time is slow and performance is low. With Glide, the performance is great, but the view creates a lot of issues the first time the app is started. 

Also, the issue is isolated to the image loading. When I disable image loading, the activity behaves as desired. 

This is the activity flow:
1) Load data from localDB
2) Load data & images into recyclerview
3) Make a network call to refresh data
4) Update localDB
5) Retrieve data from localDB and refresh recyclerview

The issue arises even when step 3-5 are not followed. I assume there is a notifyDataSetChanged sort of method that Glide calls once the image is loaded from cache which causes the view to mess up.

I have looked at the following question and it hasn't resolved the case:
https://github.com/bumptech/glide/issues/479

Loading image. This method is in the Adapter:

	
```
		if( imgSrc.matches("") != true || imgSrc != null )
		{
			Glide.with(context)
    		.load(imgSrc) 
    		//.centerCrop()
    		.into(holder.iv_profile);
			
			
			/*
			Picasso.with(context)
		   .load(imgSrc)
		   .into(holder.iv_profile);
			*/
		}

```
**ItemFriendDetail.xml**

```
<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res/com.hellboy.beat"
    android:id="@+id/container_frienddetail"
    android:layout_width="fill_parent"
    android:layout_height="fill_parent" >

    <ImageView
        android:id="@+id/iv_frienddetail_profilePic"
        android:layout_width="match_parent"
        android:layout_height="match_parent"
        android:layout_alignParentTop="true"
        android:adjustViewBounds="true"
        android:layout_centerHorizontal="true"
        android:scaleType="centerCrop" />

    <View
        android:id="@+id/view_storyCard_gradient"
        android:layout_width="match_parent"
        android:layout_height="match_parent"
        android:background="@drawable/bg_profile_story_gradient" />

    <LinearLayout
        android:layout_width="wrap_content"
        android:layout_height="wrap_content"
        android:layout_alignParentBottom="true"
        android:layout_marginBottom="8dp"
        android:layout_marginLeft="8dp"
        android:layout_marginRight="8dp"
        android:orientation="vertical" >

        <TextView
            android:id="@+id/tv_title"
            android:layout_width="wrap_content"
            android:layout_height="wrap_content"
            android:layout_alignTop="@+id/iv_frienddetail_profilePic"
            android:layout_toRightOf="@+id/iv_frienddetail_profilePic"
            android:alpha="0.9"
            android:ellipsize="end"
            android:maxLines="2"
            android:textColor="@color/white"
            android:textSize="16sp" />

        <TextView
            android:id="@+id/tv_cannon"
            android:layout_width="wrap_content"
            android:layout_height="wrap_content"
            android:alpha="0.5"
            android:ellipsize="end"
            android:maxLines="1"
            android:text="TextView"
            android:textColor="@color/white"
            android:textSize="12sp" />
    </LinearLayout>

</RelativeLayout>
```

Entire Adapter Method: http://pastebin.com/H0vUvcWX

Error Video: https://drive.google.com/open?id=0Bx-PuSeVhh-oRDlGWHNEaUJlYUk

**Comment by TWiStErRob** [comment]
There are a few things stand out, some are probably not related:

 1. Move your span calculations and `setLayoutParams` before the Glide call. That may confuse Glide.
 2. `imgSrc.matches("") != true || imgSrc != null` is always true (`""` is non null and `matches` would throw NPE if it was ever null) and inefficient (matches compiles a regex pattern), use `imgSrc != null && ! "".equals(imgSrc)`
 3. The same `if` is missing an `else`, when the holder is recycled you may need to clear the image: `... else { Glide.clear(holder.iv_profile); }`.
 4. `position` is not final, move the `setOnClickListener` to `onCreateViewHolder` or `SimpleViewHolder` constructor and use `mItems.get(holder.getAdapterPosition())`. Regarding this last one: I suggest you don't touch a single line code more until you've seen this video: https://youtu.be/imsr8NrIAMs?list=PLWz5rJ2EKKc_Tt7q77qwyKRgytF1RzRx8; it was the best investment of time I had watching a video in a long time.

Let me know if it still reproes after fixing 1).

**Comment by summers314** [comment]
Hi. Thanks a lot for taking the time out to go through the code and suggesting improvements. 

I've implemented points 1-3 .

I've moved the Glide call to the end of the onBindViewHolder method. So the entire adapter class is mostly the same , except this part (snippet): 
```
...
if (lp.rowSpan != rowSpan || lp.colSpan != colSpan) 
        {
            lp.rowSpan = rowSpan;
            lp.colSpan = colSpan;
            itemView.setLayoutParams(lp);
        }
        
        if( imgSrc != null && !"".equals(imgSrc) )
		{
			Glide.with(context)
    		.load(imgSrc) 
    		.into(holder.iv_profile);	
		}else
		{
			Glide.clear(holder.iv_profile);
		}
```

Further. The issue occurs even when I wait for a few seconds after the activity has begun and scroll the list slowly. Here is how the recyclerView is created in the parent activity. (However I doubt this is the issue,since the scrolling is smooth when I disable imageLoading)

```
		mRecyclerView = (TwoWayView) findViewById(R.id.listView_friendDetail);
		mRecyclerView.setHasFixedSize(true);
		mRecyclerView.setLongClickable(true);
		
        final Drawable divider = getResources().getDrawable(R.drawable.divider);
        mRecyclerView.addItemDecoration(new DividerItemDecoration(divider));
		
		adapter_details = new Adapter_FriendDetail(Activity_frienddetail.this, mRecyclerView, fData_list);
		adapter_details.setCallback(this);
		
		mRecyclerView.setAdapter(adapter_details);
		
		mRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() 
		{
			@Override
		    public void onScrolled(RecyclerView recyclerView, int dx, int dy) 
		    {
				
		    }
		});

```
I have tried various permutations with `mRecyclerView.setHasFixedSize()` set to `true/false`. Also with & without `mRecyclerView.addItemDecoration(new DividerItemDecoration(divider))` , but the issue remains

**Comment by summers314** [comment]
It almost feels like my RecyclerView is another ScrollView and the scrolling is leading the RecyclerView to scroll off the screen. (Even though the recyclverView is inside a RelativeLayout)

**Comment by TWiStErRob** [comment]
What does your scroll listener do? Did you empty it for posting it here, or is it really just an empty shell?

Does your `mItems` change? Try logging/breaking on those lines. Btw, you need to notifyBlahBlah() right after you make those changes.

For me it looks like there's something added at the top that pushes the view down. Like a sticky header or something. What happens if you use a constant span sizes for all of the items? Is that the correct way of setting the size? It feels really hacky, because the adapter (dealing with data binding) has to know about which manager the list is using `GridLayoutManager` has `SpanSizeLookup`.

**Comment by summers314** [comment]
1. Yes the scroll listener is just empty. 

2. Yes it does change after the data is retrieved from the cloud.  But even when I'm not doing it, the error remains.  I've commented out just the glide part, and the view doesn't give the error. As you can see in the video. The problem occurs only the first time the images are loaded from the cache. After that when I come back to this activity, it doesn't happen anymore. Infact, after that I can make network calls and update the view, and even then the error doesn't occur.

3. It occurs even when I comment out the span part from the adapter, and use constant sizes. I'm using the following library, and according to it the usage is correct: 
Android - Two Way View (Recycler View)
https://github.com/lucasr/twoway-view

There is some sort of updateView-method that Glide is calling after the image is ready to load, that leads to this.

**Comment by TWiStErRob** [comment]
> There is some sort of updateView-method that Glide is calling after the image is ready to load

Yes there is, it's called `ImageView.setImageDrawable`.

> How Glide refreshes the view differently

It is not, the only difference is that a memory cache hit (2nd load onwards) is called sync, so by the time onBind returns all the images are already set in the list items. This makes it easier for the layout manager to do the layout. I guess it doesn't like when Glide asynchronously sets the image, which is weird because Picasso should be doing that too. Does it make any difference if you set `android:adjustViewBounds="false"`?

Hmm... I seem to remember something (though I may be wrong): Picasso may have cross-fade off by default, while Glide has it on. This means that `.dontAnimate()` should better match Picasso's behavior. Aside from the sync `setImageDrawable` call, the memory cache hits also don't have animations.

**Comment by summers314** [comment]
Correction: With larger data-sets even Picasso gives the same error. Also, previously when I went back and restarted the activity the second time the images loaded without a glitch. But with a larger set (20-items), the glitch occurs everytime (even when loading images from cache). 

No difference when I set `android:adjustViewBounds="false"`

Has this been reported with people using RecyclerView? Its possible there is a issue with the RecyclerView library that I'm using. Its not compatible with Glide.  By the behaviour I think its something in the onScroll method of the library. 

Also, I've seen a lot of questions mention setSpan / setGapStrategy methods; which in case of this library are not accessible.
